src/mito2/src/sst/parquet/reader.rs

```diff
@@ -515,6 +552,100 @@ impl ParquetReader {
         Ok(None)
     }
 
+    /// Prunes batches by the pushed down predicate.
+    fn prune_batches(&mut self) -> Result<()> {
+        // fast path
+        if self.predicate.is_empty() {
+            return Ok(());
+        }
+
+        let mut new_batches = VecDeque::new();
+        let mut batches = std::mem::take(&mut self.batches);
+        for batch in batches.drain(..) {
+            let Some(batch_filtered) = self.precise_filter(batch)? else {
+                continue;
+            };
+            if !batch_filtered.is_empty() {
+                new_batches.push_back(batch_filtered);
+            }
+        }
+        self.batches = new_batches;
+
+        Ok(())
+    }
+
+    /// TRY THE BEST to perform pushed down predicate precisely on the input batch.
+    ///
+    /// Supported filter expr type is defined in [SimpleFilterEvaluator].
+    ///
+    /// When a filter is referencing primary key column, this method will decode
+    /// the primary key and put it into the batch.
+    fn precise_filter(&self, mut input: Batch) -> Result<Option<Batch>> {
+        let mut mask = BooleanBuffer::new_set(input.num_rows());
+
+        let field_id_map = self
+            .read_format
+            .metadata()
+            .field_columns()
+            .enumerate()
+            .map(|(i, f)| (f.column_id, i))
+            .collect::<HashMap<_, _>>();
+
+        // Run filter one by one and combine them result
+        // TODO(ruihang): run primary key filter first. It may short circuit other filters
+        for filter in &self.predicate {
```

Comment:
if the mask is all false, we can break it early

Reply:
We don't have a cheap way to check this :smiling_face_with_tear: